File: mobile/fill_translations.py
#!/usr/bin/env python3
"""Fill all 146 i18n keys in every language file with Hindi (English) source values.
This ensures every screen label is present in all 22 languages.
Team can then replace English values with native-script translations per language."""
import json, os

HI_FILE = './mobile/public/locales/hi/common.json'
LANGS_DIR = './mobile/public/locales'
LANGS = ['as','bn','brx','doi','gu','hi','kn','ks','kok','mai','ml','mni','mr','ne','or','pa','sa','sat','sd','ta','te','ur']

hi = json.load(open(HI_FILE))
used_keys = sorted(k.strip() for k in open('/tmp/i18n_keys.txt')
                   if '.' in k.strip() and not k.startswith('/'))
hi_keys_in_use = [k for k in used_keys if k in hi]

print(f'Keys to fill: {len(hi_keys_in_use)}')

for lang in LANGS:
    path = f'{LANGS_DIR}/{lang}/common.json'
    lang_data = json.load(open(path))

    filled = 0
    for key in hi_keys_in_use:
        if key not in lang_data:
            lang_data[key] = hi[key]
            filled += 1

    with open(path, 'w', encoding='utf-8') as f:
        json.dump(lang_data, f, ensure_ascii=False, indent=2)

    print(f'{lang}: {filled} missing keys filled, file has {len(lang_data)} total keys')

print('\nDone. Regenerate resources: node mobile/scripts/generate-i18n-resources.js')
print('Then commit and push.')

File: mobile/scripts/fill_translations.ts
#!/usr/bin/env node
/**
 * fill_translations.ts
 * Ensures every i18n key used in source files is present in all 22 language files.
 * Reads keys from TypeScript source directly — no shell dependencies.
 * Missing keys get the Hindi (English) source value; existing translations are preserved.
 */
import * as fs from 'fs';
import * as path from 'path';

const LANGS_DIR = path.resolve(__dirname, '..', 'public', 'locales');
const HI_FILE = path.resolve(__dirname, '..', 'public', 'locales', 'hi', 'common.json');
const LANGS = [
  'as','bn','brx','doi','gu','hi','kn','ks','kok',
  'mai','ml','mni','mr','ne','or','pa','sa','sat','sd','ta','te','ur',
];

// Recursively find all .ts/.tsx files
function getSourceFiles(dir: string): string[] {
  const entries = fs.readdirSync(dir, { withFileTypes: true });
  const files: string[] = [];
  for (const e of entries) {
    const full = path.join(dir, e.name);
    if (e.isDirectory()) {
      files.push(...getSourceFiles(full));
    } else if (/\.t(sx)?$/.test(e.name)) {
      files.push(full);
    }
  }
  return files;
}

// Extract i18n keys from source content
function extractKeys(src: string): string[] {
  const keys: string[] = [];
  // i18n keys only contain: letters, digits, underscore, hyphen, dot
  for (const m of src.matchAll(/t\(['"]([a-zA-Z0-9_.-]+)['"]\)/g)) {
    keys.push(m[1]);
  }
  return keys;
}

const srcFiles = getSourceFiles(path.resolve(__dirname, '..', 'src'));
const keySet = new Set<string>();
for (const f of srcFiles) {
  const src = fs.readFileSync(f, 'utf-8');
  for (const k of extractKeys(src)) keySet.add(k);
}
const hiKeysInUse = [...keySet].sort();

console.log(`Keys to fill: ${hiKeysInUse.length}`);

// Load Hindi (source — all English values)
const hi: Record<string, string> = JSON.parse(fs.readFileSync(HI_FILE, 'utf-8'));

let grandFilled = 0;

for (const lang of LANGS) {
  const filePath = path.join(LANGS_DIR, lang, 'common.json');
  const langData: Record<string, unknown> = JSON.parse(fs.readFileSync(filePath, 'utf-8'));

  let filled = 0;
  for (const key of hiKeysInUse) {
    if (!(key in langData)) {
      langData[key] = hi[key] ?? key;
      filled++;
    }
  }

  fs.writeFileSync(filePath, JSON.stringify(langData, null, 2), 'utf-8');
  grandFilled += filled;
  console.log(`${lang}: filled=${filled}, total=${hiKeysInUse.length}`);
}

console.log(`\nTotal missing keys filled: ${grandFilled}`);
console.log('Regenerate resources: node scripts/generate-i18n-resources.js');